crates/bevy_ui/src/lib.rs

```diff
@@ -178,72 +196,169 @@ impl Plugin for UiPlugin {
-                ui_focus_system.in_set(UiSystem::Focus).after(InputSystem),
             );
-
-        let ui_layout_system_config = ui_layout_system
-            .in_set(UiSystem::Layout)
-            .before(TransformSystem::TransformPropagate);
-
-        let ui_layout_system_config = ui_layout_system_config
-            // Text and Text2D operate on disjoint sets of entities
-            .ambiguous_with(bevy_text::update_text2d_layout)
-            .ambiguous_with(bevy_text::detect_text_needs_rerender::<bevy_text::Text2d>);
-
+        #[cfg(feature = "bevy_render")]
         app.add_systems(
-            PostUpdate,
-            (
-                check_visibility::<With<Node>>.in_set(VisibilitySystems::CheckVisibility),
-                update_target_camera_system.in_set(UiSystem::Prepare),
-                ui_layout_system_config,
-                ui_stack_system
-                    .in_set(UiSystem::Stack)
-                    // the systems don't care about stack index
-                    .ambiguous_with(update_clipping_system)
-                    .ambiguous_with(ui_layout_system)
-                    .in_set(AmbiguousWithTextSystem),
-                update_clipping_system.after(TransformSystem::TransformPropagate),
-                // Potential conflicts: `Assets<Image>`
-                // They run independently since `widget::image_node_system` will only ever observe
-                // its own ImageNode, and `widget::text_system` & `bevy_text::update_text2d_layout`
-                // will never modify a pre-existing `Image` asset.
-                widget::update_image_content_size_system
-                    .in_set(UiSystem::Prepare)
-                    .in_set(AmbiguousWithTextSystem)
-                    .in_set(AmbiguousWithUpdateText2DLayout),
-            ),
+            PreUpdate,
+            ui_focus_system.in_set(UiSystem::Focus).after(InputSystem),
         );
+
+        let systems = make_systems();
+        app.add_systems(PostUpdate, systems);
+
+        #[cfg(feature = "bevy_text")]
         build_text_interop(app);
 
         #[cfg(feature = "bevy_ui_picking_backend")]
         app.add_plugins(picking_backend::UiPickingPlugin);
 
+        #[cfg(feature = "bevy_render")]
         if !self.enable_rendering {
             return;
         }
 
+        #[cfg(feature = "bevy_render")]
         build_ui_render(app);
     }
 
-    fn finish(&self, app: &mut App) {
+    fn finish(&self, _app: &mut App) {
+        #[cfg(feature = "bevy_render")]
         if !self.enable_rendering {
             return;
         }
 
-        let Some(render_app) = app.get_sub_app_mut(RenderApp) else {
+        #[cfg(feature = "bevy_render")]
+        let Some(render_app) = _app.get_sub_app_mut(RenderApp) else {
             return;
         };
 
+        #[cfg(feature = "bevy_render")]
         render_app.init_resource::<UiPipeline>();
     }
 }
 
+#[cfg(all(feature = "bevy_text", feature = "bevy_render"))]
+fn make_systems() -> (
```

Comment:
I really dislike this approach; it's very complex.

Reply:
What is a better approach?

Reply:
Put each call to add_systems that needs to be conditional on its own line, with the appropriate cfg flag attached.